successX now correctly shows html output for `nim doc`, `nim jsondoc`; fix #13121

timotheecour opened the pull request:
* also, fixes https://github.com/nim-lang/Nim/issues/13121

timotheecour committed:
successX now correctly shows html output for nim doc

Araq commented on compiler/docgen.nim:
```diff
@@ -39,7 +39,7 @@ type
     emitted: IntSet # we need to track which symbols have been emitted
                     # already. See bug #3655
-    destFile*: AbsoluteFile
+    destFile*: AbsoluteFile # that's a lie...
```
Well ok, I believe you. Can we fix it so that it doesn't lie?

timotheecour replied:
done (https://github.com/nim-lang/Nim/issues/13121 was root caused; this PR includes a fix for it)

timotheecour committed:
_

timotheecour committed:
fix #13121

timotheecour committed:
fixup hintSuccessX to be less weird

timotheecour commented:
PTAL.
* I've also fixed https://github.com/nim-lang/Nim/issues/13121 to address above comment
* i added semi-unrelated change:
```
Hint: 53431 LOC; 0.131 sec; 46.711MiB peakmem; Debug build; /patho/thello.nim proj; /patho/thello.html out [SuccessX]                                                    
=>
Hint: 53431 LOC; 0.131 sec; 46.711MiB peakmem; Debug build; proj: /patho/thello.nim; out: /patho/thello.html [SuccessX]
```

which is less weird looking

Araq commented on compiler/options.nim:
```diff
@@ -244,6 +244,7 @@ type
     searchPaths*: seq[AbsoluteDir]
     lazyPaths*: seq[AbsoluteDir]
     outFile*: RelativeFile
+    outFileAlt*: AbsoluteFile ## eg html produced by `nim doc`
```
Why do we need this?

timotheecour replied:
fixed; now re-using `outFile` instead and it works

Araq commented on compiler/main.nim:
```diff
@@ -363,8 +363,11 @@ proc mainCommand*(graph: ModuleGraph) =
-    let output = if optListFullPaths in conf.globalOptions: $conf.getOutFileFull else: $conf.outFile
+    template postprocess(path): untyped =
+      if optListFullPaths in conf.globalOptions: $path
+      else: path.`$`.lastPathPart
```
I don't understand this.

timotheecour replied:
you had commented here https://github.com/nim-lang/Nim/pull/13043#issuecomment-570941009
> I see the value in it but omg it's verbose

and this allows shortening things when `--listfullpaths:off` (as was done here https://github.com/nim-lang/Nim/pull/13043)

that's the simplest code I can think of that does what we want

Araq replied:
``lastPathPart`` is never what I want as my filenames do not in end in slashes anyway. The old code

```nim

let project = if optListFullPaths in conf.globalOptions: $conf.projectFull else: $conf.projectName
let output = if optListFullPaths in conf.globalOptions: $conf.getOutFileFull else: $conf.outFile

```

was perfectly clear code. Clearer than your replacement which I still don't understand.

timotheecour replied:
old code was incorrect because of the logic inside `absOutFile` which sometimes adds `.out` suffix and wasn't reflected here; I've now fixed it differently without using lastPathPart and it should work in all cases

Araq commented on compiler/docgen.nim:
```diff
@@ -1054,12 +1054,13 @@ proc writeOutput*(d: PDoc, useWarning = false) =
   if optStdout in d.conf.globalOptions:
     writeRope(stdout, content)
   else:
-    template outfile: untyped = d.destFile
```
But this style avoided a copy.

timotheecour replied:
IIUC it's guaranteed to be a move in this case; for eg see below:
```nim
type Foo = distinct string
type Bar = ref object
  foo: Foo

proc main(a: Bar)=
  let b = a.foo
  doAssert cast[int](a.foo.string[0].unsafeAddr) == cast[int](b.string[0].unsafeAddr)

let a = Bar(foo: "abc".Foo)
main(a)
```

if i added `doAssert cast[int](outfile.string[0].unsafeAddr) == cast[int](d.destFile.string[0].unsafeAddr)` below `let outfile = d.destFile`, the assert would also pass. 

`let` is cleaner in this case

Araq replied:
That's a bug, finally addressed in --gc:arc, the move is unsound.

timotheecour replied:
* but `--gc:arc` also makes the assert pass ; let's discuss this somewhere dedicated to this instead of here ; EDIT :here: https://github.com/nim-lang/Nim/issues/13140
* I've changed it back to `template outfile: untyped = d.destFile` to unblock this PR;  `byRef: outfile = d.destFile` would be what I'd use once https://github.com/nim-lang/Nim/pull/11824 is merged; and it's also shorter than `template outfile: untyped = d.destFile`

timotheecour committed:
remove outFileAlt

timotheecour committed:
address comment

timotheecour commented:
PTAL

Araq commented on compiler/pathutils.nim:
```diff
@@ -26,6 +26,14 @@ proc copyFile*(source, dest: AbsoluteFile) =
+template toRelative(x: typedesc[AnyPath]): typedesc =
+  when x is AbsoluteFile | RelativeFile: RelativeFile else: RelativeDir
+
+proc lastPathPart*(x: AnyPath): auto =
```
The following rule served me well: If you cannot write down the return type what you wrote is not a ``proc`` or ``func``, but something else.

timotheecour replied:
I've removed `lastPathPart`

Araq commented on compiler/pathutils.nim:
```diff
@@ -54,21 +62,21 @@ when true:
 
   proc `==`*[T: AnyPath](x, y: T): bool = eqImpl(x.string, y.string)
 
-  proc `/`*(base: AbsoluteDir; f: RelativeFile): AbsoluteFile =
```
In my world bugfixes are not a justification to rewrite code in a style you happen to like better.

timotheecour replied:
reverted to the previous way (which had a lot of code duplication) and instead using a `postProcessBase` which handles the empty `base` case; the correct behavior would be to throw when base is `empty` as was ruled here https://github.com/nim-lang/Nim/pull/10018#issuecomment-448192956 however this would require more unrelated changes and should be done in separate PR; this PR doesn't change externally visible behavior but at least doesn't "lie" (see https://github.com/nim-lang/Nim/pull/13116#discussion_r365580322)

timotheecour committed:
bugfix: use absOutFile instead of getOutFileFull (because of foo.out special case)

timotheecour committed:
fixup

timotheecour commented:
PTAL
